api/server/httputils/errors.go

```diff
@@ -37,49 +33,45 @@ func GetHTTPErrorStatusCode(err error) int {
 	}
 
 	var statusCode int
-	errMsg := err.Error()
 
-	switch e := err.(type) {
-	case httpStatusError:
-		statusCode = e.HTTPErrorStatusCode()
-	case inputValidationError:
+	// Stop right there
+	// Are you sure you should be adding a new error class here? Do one of the existing ones work?
+
+	// Note that the below functions are already checking the error causal chain for matches.
+	switch {
```

Comment:
It looks pretty close to 1:1. Classes of errors is a good strategy, but I think there will always be cases where the request context (url, method, etc) change the response code.

---

How would we support more detailed/structured API errors using this scheme? For example, a `NotFound` on a service  create could mean many different things (container, network, secret, etc not found).

A better api would return something like this as the body, along with the 404 error code:

```json
{
   "resource": "network",
   "id": "something",
   "msg": "Network something not found",
   ...
}
```

That way API consumers have more options to communicate the error to users.

Translating errors in the API handlers/router also makes the code much easier to read. You can identify error cases from a single place instead of digging through the entire call stack.
---

Anyway, this is still a huge improvement.